File: OTCDerivativesCalculatorModule/Project_Cpp/lib_static/FpmlSerialized/GenClass/fpml-com-5-4/CommoditySpread.cpp
// CommoditySpread.cpp 
#include "CommoditySpread.hpp"
#ifdef ConsolePrint
    #include <iostream>
#endif 

namespace FpmlSerialized {

CommoditySpread::CommoditySpread(TiXmlNode* xmlNode)
: Money(xmlNode)
{
    #ifdef ConsolePrint
        std::string initialtap_ = FileManager::instance().tap_;
        FileManager::instance().tap_.append("   ");
    #endif 
   //spreadConversionFactorNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* spreadConversionFactorNode = xmlNode->FirstChildElement("spreadConversionFactor");

   if(spreadConversionFactorNode){spreadConversionFactorIsNull_ = false;}
   else{spreadConversionFactorIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- spreadConversionFactorNode , address : " << spreadConversionFactorNode << std::endl;
   #endif
   if(spreadConversionFactorNode)
   {
      if(spreadConversionFactorNode->Attribute("href") || spreadConversionFactorNode->Attribute("id"))
      {
          if(spreadConversionFactorNode->Attribute("id"))
          {
             spreadConversionFactorIDRef_ = spreadConversionFactorNode->Attribute("id");
             spreadConversionFactor_ = boost::shared_ptr<XsdTypeDecimal>(new XsdTypeDecimal(spreadConversionFactorNode));
             spreadConversionFactor_->setID(spreadConversionFactorIDRef_);
             IDManager::instance().setID(spreadConversionFactorIDRef_,spreadConversionFactor_);
          }
          else if(spreadConversionFactorNode->Attribute("href")) { spreadConversionFactorIDRef_ = spreadConversionFactorNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { spreadConversionFactor_ = boost::shared_ptr<XsdTypeDecimal>(new XsdTypeDecimal(spreadConversionFactorNode));}
   }

   //spreadUnitNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* spreadUnitNode = xmlNode->FirstChildElement("spreadUnit");

   if(spreadUnitNode){spreadUnitIsNull_ = false;}
   else{spreadUnitIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- spreadUnitNode , address : " << spreadUnitNode << std::endl;
   #endif
   if(spreadUnitNode)
   {
      if(spreadUnitNode->Attribute("href") || spreadUnitNode->Attribute("id"))
      {
          if(spreadUnitNode->Attribute("id"))
          {
             spreadUnitIDRef_ = spreadUnitNode->Attribute("id");
             spreadUnit_ = boost::shared_ptr<QuantityUnit>(new QuantityUnit(spreadUnitNode));
             spreadUnit_->setID(spreadUnitIDRef_);
             IDManager::instance().setID(spreadUnitIDRef_,spreadUnit_);
          }
          else if(spreadUnitNode->Attribute("href")) { spreadUnitIDRef_ = spreadUnitNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { spreadUnit_ = boost::shared_ptr<QuantityUnit>(new QuantityUnit(spreadUnitNode));}
   }

    #ifdef ConsolePrint
        FileManager::instance().tap_ = initialtap_;
    #endif 
}
boost::shared_ptr<XsdTypeDecimal> CommoditySpread::getSpreadConversionFactor()
{
   if(!this->spreadConversionFactorIsNull_){
        if(spreadConversionFactorIDRef_ != ""){
             return boost::shared_static_cast<XsdTypeDecimal>(IDManager::instance().getID(spreadConversionFactorIDRef_));
        }else{
             return this->spreadConversionFactor_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<XsdTypeDecimal>();
   }
}
boost::shared_ptr<QuantityUnit> CommoditySpread::getSpreadUnit()
{
   if(!this->spreadUnitIsNull_){
        if(spreadUnitIDRef_ != ""){
             return boost::shared_static_cast<QuantityUnit>(IDManager::instance().getID(spreadUnitIDRef_));
        }else{
             return this->spreadUnit_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<QuantityUnit>();
   }
}
}

File: OTCDerivativesCalculatorModule/Project_CSharp/XmlBuilder/XmlSerializedClass/fpml-com-5-4/CommoditySpread.cs
using System;
using System.Text;
using System.Collections.Generic;
using System.Xml;

namespace FpmlSerializedCSharp
{
    public class CommoditySpread : Money
    {
        public CommoditySpread(XmlNode xmlNode)
        : base(xmlNode)
        {
            XmlNode spreadConversionFactorNode = xmlNode.SelectSingleNode("spreadConversionFactor");
            
            if (spreadConversionFactorNode != null)
            {
                if (spreadConversionFactorNode.Attributes["href"] != null || spreadConversionFactorNode.Attributes["id"] != null) 
                {
                    if (spreadConversionFactorNode.Attributes["id"] != null) 
                    {
                        spreadConversionFactorIDRef_ = spreadConversionFactorNode.Attributes["id"].Value;
                        XsdTypeDecimal ob = new XsdTypeDecimal(spreadConversionFactorNode);
                        IDManager.SetID(spreadConversionFactorIDRef_, ob);
                    }
                    else if (spreadConversionFactorNode.Attributes["href"] != null)
                    {
                        spreadConversionFactorIDRef_ = spreadConversionFactorNode.Attributes["href"].Value;
                    }
                    else
                    {
                        spreadConversionFactor_ = new XsdTypeDecimal(spreadConversionFactorNode);
                    }
                }
                else
                {
                    spreadConversionFactor_ = new XsdTypeDecimal(spreadConversionFactorNode);
                }
            }
            
        
            XmlNode spreadUnitNode = xmlNode.SelectSingleNode("spreadUnit");
            
            if (spreadUnitNode != null)
            {
                if (spreadUnitNode.Attributes["href"] != null || spreadUnitNode.Attributes["id"] != null) 
                {
                    if (spreadUnitNode.Attributes["id"] != null) 
                    {
                        spreadUnitIDRef_ = spreadUnitNode.Attributes["id"].Value;
                        QuantityUnit ob = new QuantityUnit(spreadUnitNode);
                        IDManager.SetID(spreadUnitIDRef_, ob);
                    }
                    else if (spreadUnitNode.Attributes["href"] != null)
                    {
                        spreadUnitIDRef_ = spreadUnitNode.Attributes["href"].Value;
                    }
                    else
                    {
                        spreadUnit_ = new QuantityUnit(spreadUnitNode);
                    }
                }
                else
                {
                    spreadUnit_ = new QuantityUnit(spreadUnitNode);
                }
            }
            
        
        }
        
    
        #region SpreadConversionFactor_
        private XsdTypeDecimal spreadConversionFactor_;
        public XsdTypeDecimal SpreadConversionFactor_
        {
            get
            {
                if (this.spreadConversionFactor_ != null)
                {
                    return this.spreadConversionFactor_; 
                }
                else if (this.spreadConversionFactorIDRef_ != null)
                {
                    spreadConversionFactor_ = IDManager.getID(spreadConversionFactorIDRef_) as XsdTypeDecimal;
                    return this.spreadConversionFactor_; 
                }
                else
                {
                      return this.spreadConversionFactor_; 
                }
            }
            set
            {
                if (this.spreadConversionFactor_ != value)
                {
                    this.spreadConversionFactor_ = value;
                }
            }
        }
        #endregion
        
        public string spreadConversionFactorIDRef_ { get; set; }
        #region SpreadUnit_
        private QuantityUnit spreadUnit_;
        public QuantityUnit SpreadUnit_
        {
            get
            {
                if (this.spreadUnit_ != null)
                {
                    return this.spreadUnit_; 
                }
                else if (this.spreadUnitIDRef_ != null)
                {
                    spreadUnit_ = IDManager.getID(spreadUnitIDRef_) as QuantityUnit;
                    return this.spreadUnit_; 
                }
                else
                {
                      return this.spreadUnit_; 
                }
            }
            set
            {
                if (this.spreadUnit_ != value)
                {
                    this.spreadUnit_ = value;
                }
            }
        }
        #endregion
        
        public string spreadUnitIDRef_ { get; set; }
        
    
        
    
    }
    
}

